2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,476 @@
+
+    As a PowerShell user,
+    I can easily install modules without specifying lots of switches,
+    so that I can be more productive.
+
+    As a PowerShellGet contributor,
+    I can more easily contribute code,
+    so that PowerShellGet evolves more quickly.
+
+## User Experience
+
+```powershell
+PS> Install-PSResource HappyDays
+```
+
+## Specification
+
+### Rewrite of PowerShellGet
+
+PowerShellGet is currently written as PowerShell script with a dependency on PackageManagement.
+Proposal is to write PSResource in C# to reduce complexity and make easier to maintain.
+In addition, remove dependency on PackageManagement completely as well as dependency on
+nuget.exe.
+This module will depend on Nuget Server and Client packages including https://www.nuget.org/packages/NuGet.Client.
+This module would be shipped in the PSGallery supporting PowerShell 5.1+.
+This module will likely ship in PowerShell 7.1 alongside PowerShellGet 2.x.
+
+### Side-by-side with PowerShellGet
+
+Since the current PowerShellGet 2.x version is a script module and this new one
+is a C# based module, they can coexist side-by-side.
+
+Script and module metadata will retain compatible formatting as it exists with v2.
+
+We will introduce an `Enable-PowerShellGetAlias` cmdlet that will allow users to use known cmdlets like Install-Module
+from the PowerShellGet 3.0 implementation.
+This additional cmdlet may not ship in PowerShellGet 3.0 but is planned as a part of the transition effort.
+
+### Local cache
+
+Instead of always connecting to a repository to perform an online search,
+`Find-PSResource` (see below) works against a local cache.
+This will also enable changes in PowerShell to use `Find-PSResource -Type Command` to look
+in this cache when it can't find a command and suggest to the user the module to install to
+get that command.
+This will be a local json file (one per repository) containing sufficient metadata
+for searching for resources and their dependencies.
+The cache will be stored in the user path.
+There is no system cache that is shared.
+A system wide cache would require elevation or sudo to create and update preventing
+it from being useful.
+
+Example cache entry:
+
+```json
+{
+  "name": "This is my module",
+  "exportedFunctions": [
+    "Get-One",
+    "Set-Two"
+  ],
+  "version": "1.0.0.0",
+  "type": "module",
+  "tags": [
+    "Linux",
+    "PSEdition_Core",
+    "PSEdition_Desktop",
+    "AzureAutomation"
+  ]
+}
+```
+
+An example cache with 5000 resources (approximately the number of unique packages
+currently published to PowerShellGallery.com) is ~700KB in compressed json form.
+
+The cache would have both latest stable and latest prerelease versions of resources.
+
+>[!NOTE]
+>Need to experiment if it makes sense to have a single cache file for all repositories
+>or a different cache per repository for size and perf reasons.
+>Perf tests will determine if the cache needs to be in binary form.

```

Comment:
Also & I just finished testing this, to get a single output from the Gallery now of all the Modules & versions took over 2hours and at well over 5.5k individual calls of Find-Module to the gallery to get all the info required for all modules in the Gallery, so therefore it would be good if we could get the option for both the slim file with the suggestions above, and the full fat file with everything as a choice.